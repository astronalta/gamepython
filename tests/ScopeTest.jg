# jgtest exec on

Val < Value {
    test String
    int var Int
    @init(ptest String, pint Int) {
        test = ptest
        int = pint
        Io::println("init #{self.test} #{self.int}") 
    }
    @destroy() {
        Io::println("destroy #{self.test} #{self.int}") 
    }
    str?() String { ret self.int.str }
}

Obj < Object {
    test String
    int Int
    @init(ptest String, pint Int) {
        test = ptest
        int = pint
        Io::println("init #{self.test} #{self.int}") 
    }
    @destroy() {
        Io::println("destroy #{self.test} #{self.int}") 
    }
    str?() String { ret self.int.str }
}

val(test String, int Int) Val {
    ret Val(test, int)
}

obj(test String, int Int) Obj {
    ret Obj(test, int)
}

val_cond_no_assign() {
    if true {
        Val("val_cond_no_assign", 1)            
    }
    Io::println("break")
    if true {
        val("val_cond_no_assign", 2)
    }
}

val_cond_assign() {
    x = if true {
        Val("val_cond_assign", 1)            
    } else {
        Val("val_cond_assign", 2)
    }
    Io::println("break")
    y = if true {
        val("val_cond_assign", 3)
    } else {
        Val("val_cond_assign", 4)
    }
}

val_end() {
    Val("val_end", 1)
}

val_return() Val {
    ret Val("val_return", 1)
}

val_loop_no_assign() { # FIXME: Broken (doesn't free inside loop)
    i var = 0
    while i < 2 {
        ++i
        Val("val_loop_no_assign", i)
    }
}

#val_loop_assign() { # FIXME: Broken (doesn't free inside loop)
#    i var = 0
#    y = while i < 2 {
#        ++i
#        Val("val_loop_assign", i)
#    }
#}

val_nested_no_assign() {
    i var = 0
    while i < 2 {
        ++i
        if true {
            Val("val_nested_no_assign", i)
        } 
    }
}

val_nested_assign() {
    y = if true {
        if true {
            Val("val_nested_assign", 0)
        } else {
            Val("val_nested_assign", 1)
        } 
    } else {
        Val("val_nested_assign", 2)
    }
    Io::println("break")
}

val_nested_double_assign() { # FIXME: 'x' shouldn't be freed
    y = if true {
        x = if true {
            Val("val_nested_double_assign", 1)
        } else {
            Val("val_nested_double_assign", 2)
        } 
    } else {
        Val("val_nested_double_assign", 3)
    }
    Io::println("break")
}

val_cond_var_no_assign() {
    if true {
        x = Val("val_cond_var_no_assign", 1)
        x
    }
}

val_cond_var_assign() {
    y = if true {
        x = Val("val_cond_var_assign", 1)
        x
    } else {
        Val("val_cond_var_assign", 2)
    }
    y.int = 2
}

obj_nested_double_assign() { 
    y = if true {
        x = if true {
            Obj("obj_nested_double_assign", 1)
        } else {
            nil
        }
    } else {
        nil
    }
    Io::println(y.test.str)
}

main() Int {
    val_cond_no_assign()
    Io::println('')
    val_cond_assign()
    Io::println('')
    val_loop_no_assign()
    Io::println('')
    #val_loop_assign()
    val_end()
    Io::println('')
    val_return()
    Io::println('')
    val_nested_no_assign()
    Io::println('')
    val_nested_assign()
    Io::println('')
    val_nested_double_assign()
    Io::println('')
    val_cond_var_no_assign()
    Io::println('')
    val_cond_var_assign()
    Io::println('')
    obj_nested_double_assign()
    ret 0
}
