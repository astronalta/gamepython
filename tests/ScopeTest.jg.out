init val_cond_no_assign 1
destroy val_cond_no_assign 1
break
init val_cond_no_assign 2
destroy val_cond_no_assign 2

init val_cond_assign 1
break
init val_cond_assign 3
destroy val_cond_assign 3
destroy val_cond_assign 1

init val_loop_no_assign 1
destroy val_loop_no_assign 1
init val_loop_no_assign 2
destroy val_loop_no_assign 2

init val_end 1
destroy val_end 1

init val_return 1
destroy val_return 1

init val_nested_no_assign 1
destroy val_nested_no_assign 1
init val_nested_no_assign 2
destroy val_nested_no_assign 2

init val_nested_assign 0
break
destroy val_nested_assign 0

init val_nested_double_assign 1
destroy val_nested_double_assign 1
break
destroy val_nested_double_assign 1

init val_cond_var_no_assign 1
destroy val_cond_var_no_assign 1

init val_cond_var_assign 1
destroy val_cond_var_assign 1
destroy val_cond_var_assign 2

init obj_nested_double_assign 1
obj_nested_double_assign
destroy obj_nested_double_assign 1
